portality/lib/argvalidate.py

```diff
@@ -0,0 +1,16 @@
+from portality.core import app
+
+def argvalidate(caller_name, args_and_rules, exception):
```

Comment:
Hmm. Type and None checking. What bothers me is that this is almost a language level feature. When I write something like this I usually find myself questioning whether there's something about the technical design or (not) using language features and Python programming conventions that I'm missing.

It seems it's being introduced to keep function complexity down by being able to make certain assumptions for the rest of the function after using this. Could you elaborate on the reasons for introducing it?

I can see two main uses atm: None checking and type checking.

With the None check, I tend to focus on why it could be None, if it's possible to say so in advance. In Python I'd generally deal with None with an explicit check closest to where the value is being calculated. E.g. `s = models.Suggestion.pull(suggestion_id); if not s: ...`. Do the callers of argvalidate need to interface with unpredictable, untested or legacy code of some sort?

On the type checking: is there a possibility a function expecting a models.Suggestion could get a models.Journal and not be able to deal with that only by polymorphism? If so there may be a technical design thing to look at there. My approach in Python when expecting a certain type is to just ensure that is all that ever gets passed in. As that's a hard thing to predict, I essentially rely on tests. To be fair, Python 3 does introduce type hinting, but adding that onto a dynamically typed language seems odd - so perhaps I'm just missing something about the use case.

Reply:
I introduced this bit of code for a couple of reasons:

1. To reduce code complexity, as you say.  It automates type and None checking where those things are necessary (you can choose which to check at the start of the function, and you don't have to check either).  This gets rid of lots of tedious explicit checking in the function.

2. To eliminate common bugs.  You say you prefer to use None checking near where the value is being calculated, and this is great and good when it is remembered.  In my experience many many bugs are due to that not being done.  Similarly with object types, code rarely checks we got what we expected, which is theoretically risky in a duck typed language.

To give you an interesting instance of (2) - did you know that flask's current_user session variable is not actually our Account object?  I didn't.  It turns out if you try to make some kinds of decisions with that variable, they will not work correctly, because current_user is both not the Account object and does not present itself as the Account object - it just behaves a lot like it.  Have you ever (any of us?) type checked the current_user object before using it?

So, the point is to make the opportunity for a bug in the BLL much smaller, and to clean up the code so it is more readable.  I think this is a sensible addition - I wouldn't use it everywhere, but I do want it as a gatekeeper to the BLL.